File: backtrack_recursion/216 combination sum iii.py
"""
This problem can be solved by the template easily. 

"""
class Solution(object):
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        res = []
        self.backtrack(k, 1, n, [], res)
        return res
    
    def backtrack(self, k, index, n, path, res):
        if n == 0 and k == 0:
            res.append(path)
            return
        if n < 0 or k < 0:
            return
        for i in range(index, 10):
            self.backtrack(k-1, i+1, n - i, path+[i], res)

File: backtrack_recursion/216. Combination Sum III.cpp
class Solution {
public:
    vector<vector<int>> res;
    void combination(int k, int n, int index, vector<int> path)
    {
        if(n < 0)
            return ;
        else if(n == 0 && path.size() == k)
            res.push_back(path);
        else
        {
            for(int i = index; i < 10; i++)
            {
                if(path.empty() || (!path.empty() && i > path.back()))
                {
                    path.push_back(i);
                    combination(k, n - i, index + 1, path);
                    path.pop_back();
                }                
            }
        }
    }
    
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<int> path;
        combination(k, n, 1, path);
        return res;
    }   
};
